video/out/d3d11/ra_d3d11.c

```diff
@@ -1598,7 +1598,7 @@ static void save_cached_program(struct ra *ra, struct ra_renderpass *pass,
         .comp_bytecode_len = comp_bc.len,
     };
     memcpy(header.magic, cache_magic, sizeof(header.magic));
-    strncpy(header.compiler, spirv->name, sizeof(header.compiler) - 1);
+    strncpy(header.compiler, spirv->name, sizeof(header.compiler));
```

Comment:
I'd use [`memccpy`](https://www.man7.org/linux/man-pages/man3/memccpy.3.html) and assert that truncation doesn't occur. It's more explicit that way which should avoid unnecessary misunderstanding (and warnings) in the future.

```suggestion
    char *p = memccpy(header.compiler, spirv->name, '\0', sizeof(header.compiler));
    assert(p); // truncation shouldn't occur
```

`memccpy` should also be [available on windows](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/posix-memccpy?view=msvc-170) so there's (hopefully) no portability issues.

Reply:
On second thought, since both of these are buffers of the same fixed size, why not just plain `memcpy`?

Reply:
I don't remember now, but look like I intended to change it to `memcpy`, looking at the `Copy whole buffer.` in commit. But anyway there is virtually no difference. Both are the same size, both memcpy and strcpy will work fine, of course one will stop sooner, if possible, but there is no difference.

Reply:
>  of course one will stop sooner

Just to be clear - and since a lot of people aren't aware of this - `strncpy` *does not* stop sooner. It will nul-pad the entire `dst` buffer if the `src` string is shorter (see [strncpy manpage](https://man7.org/linux/man-pages/man3/strncpy.3p.html)).

Perhaps nul-padding was intended here? If that was the case, `memccpy` would still work since the buffer was zero initialized - but plain `memcpy` might copy trailing stuff that might not be nul-padded.

But in any case, my main point is that `strncpy` semantics are weird and so any code that uses it is usually suspect (both to human readers and static analyzers). So it's better avoided in favor of more explicit functions when possible.

Reply:
Oh, I must be tired, I was talking about strcpy, while the code indeed is using strncpy. Anyway we are just dancing around to make static analyzers happy here.

We already have zero initialized dst buffer above. And src is also valid null-terminated string, initialized elsewhere. And both buffers are the same size. So the most easy case we can have.